docs/dev/topology-over-raft.md

```diff
@@ -405,3 +411,72 @@ In case of failures like timeouts, connection issues, etc., the topology
 coordinator keeps retrying. Eventually, it can give up and just move the new
 node to the `left` state, either due to a timeout or an operator intervention,
 but this is not implemented yet.
+
+# Choosing how to join a cluster
+
+A bootstrapping node needs to determine whether to use raft topology operations
+or not. In order to do so, group 0 discovery is performed which establishes
+whether the node is supposed to create a new cluster or join an existing one:
+
+- If the node creates a new cluster, it bootstraps in raft topology mode.
+- If the node joins an existing cluster, it learns about a node which is already
+  a part of the group 0. Such a node will already know whether the cluster
+  operates in raft topology mode or not. It then issues `JOIN_NODE_QUERY`
+  RPC which tells the new node how it should join the cluster.
+
+# Upgrade from legacy topology to raft-based topology
+
+Upgrading existing clusters to use raft-based topology operations is supported.
+The process is not fully automatic because it needs to be triggered by the administrator
+after making sure that no topology operations are running at the moment. This
+limitation might be lifted in the future, and the procedure will trigger
+automatically after making sure that it's safe to do so.
+
+From the admin's point of view, the steps are as follows:
+
+- Upgrade all nodes in the cluster to the newest version
+- Wait until the `SUPPORTS_CONSISTENT_TOPOLOGY_CHANGES` feature becomes enabled
+  and an appropriate message about the cluster being ready for upgrade is printed
+- Trigger the upgrade via `POST /storage_service/raft_topology/upgrade` HTTP route
+- Monitor progress of the upgrade via `GET /storage_service/raft_topology/upgrade`
+  or via observing the logs
+- After all nodes report `done` via the GET endpoint, the upgrade has fully finished
+
+The `upgrade_state` static column in `system.topology` serves the key role
+in coordinating the upgrade. It goes through the following states in the following
+order:
+
+- `not_upgraded` (or just null) - in this state, nodes wait for upgrade to be
+  started by the administrator. Invoking the `POST /storage_service/raft_topology/upgrade`
+  route changes the state from `not_upgraded` to `build_coordinator_state`
+
+After going out of the `not_upgraded` state, the current raft leader starts its
+topology coordinator fiber and coordinates the remaining steps:
+
+- `build_coordinator_state` - the topology coordinator reads the contents
```

Comment:
A global barrier at this point could also serve as a connectivity check @gleb-cloudius @piodul.

I believe we shouldn't progress upgrade beyond this step if we don't verify that Raft commands can reach every node.

This is not something that the administrator can verify easily.

Reply:
On the other hand, if the administrator is responsible for checking that every node reached `done`, if this wait "times out", they'll have to do disaster recovery anyway. So ok, as long as the admin is responsible for monitoring the state.